crates/evm/src/executor/inspector/cheatcodes/mod.rs

```diff
@@ -201,6 +203,175 @@ pub struct Cheatcodes {
     /// Breakpoints supplied by the `vm.breakpoint("<char>")` cheatcode
     /// char -> pc
     pub breakpoints: Breakpoints,
+
+    /// Track if cool cheatcode was called on each address
+    pub addresses: rHashMap<Address, AddressState>,
+    /// Track if an addresses storage slot is cool or not
+    pub address_storage: rHashMap<Address, rHashMap<U256, StorageSlotState>>,
+    /// How much gas to charge in the next step (op code) based on cool cheatcode calculations
+    pub additional_gas_next_op: u64,
+
+    pub prev_opcode: u8,
+    pub prev_opcode_gas: Gas,
+}
+
+/// Whether an Address is accessed or not
+#[derive(Clone, PartialEq, Debug)]
+pub enum AddressState {
+    /// if already accessed, then charge WARM_STORAGE_READ_COST (100)
+    Warm,
+    /// charge COLD_ACCOUNT_ACCESS_COST (2600)
+    Cool,
+}
+
+/// Whether a Storage Slot is warm or already been modified
+#[derive(Clone, PartialEq, Debug)]
+pub enum StorageSlotState {
+    /// charge extra based on SSTORE calculations
+    WarmWithSLOAD,
+    /// if SSTORE already happened, don't charge extra
+    WarmWithSSTORE,
+    /// same as if empty
+    Cool,
+}
+
+/// Function to charge extra gas per opcode based on cool cheatcode
+fn add_gas_from_cool_cheatcode<DB: DatabaseExt>(
+    state: &mut Cheatcodes,
+    interpreter: &mut Interpreter,
+    data: &mut EVMData<'_, DB>,
+) -> InstructionResult {
+    // For gas costs, see https://eips.ethereum.org/EIPS/eip-2200, https://eips.ethereum.org/EIPS/eip-2929
+
+    // if previous step added gas, add it once
+    // note that all the opcodes will already have a cost (usually 100)
+    // so adding until it hits the gas expected for a cold key/address
+    if state.additional_gas_next_op > 0 {
+        interpreter.gas.record_cost(state.additional_gas_next_op);
+        state.additional_gas_next_op = 0;
+    }
+
+    // if cool cheatcode was ever called on this address
+    let contract_address = interpreter.contract().address.to_ethers();
+    if state.addresses.get(&contract_address).is_some() {
```

Comment:
- `vm.cool` involves two parts: making the `address` itself cool, making the storage slots cool
- check that `vm.cool` was passed to this `address`
  - if so, track a few opcodes (call/balance/etc) to mark the address as warm, as well as queuing up additional gas to be added to the next opcode. a warm address is 100, a cool one is 2600, so add 2500.
- check the storage slots
  - if.an sload or sstore, mark address as warm but add different wants of extra gas depending on the situation.